test/lit/wat-kitchen-sink.wast

```diff
@@ -181,6 +185,455 @@
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (drop
+ ;; CHECK-NEXT:   (i32.const 2)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (unreachable)
+ ;; CHECK-NEXT: )
+ (func $add-unreachable-3 (result i32)
+  i32.const 1
+  i32.const 2
+  unreachable
+ )
+
+ ;; CHECK:      (func $add-twice (type $ret2) (result i32 i32)
+ ;; CHECK-NEXT:  (tuple.make
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (i32.const 1)
+ ;; CHECK-NEXT:    (i32.const 2)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (i32.const 3)
+ ;; CHECK-NEXT:    (i32.const 4)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $add-twice (type $ret2)
+  i32.const 1
+  i32.const 2
+  i32.add
+  i32.const 3
+  i32.const 4
+  i32.add
+ )
+
+ ;; CHECK:      (func $add-twice-stacky (type $ret2) (result i32 i32)
+ ;; CHECK-NEXT:  (local $scratch i32)
+ ;; CHECK-NEXT:  (tuple.make
+ ;; CHECK-NEXT:   (block (result i32)
+ ;; CHECK-NEXT:    (local.set $scratch
+ ;; CHECK-NEXT:     (i32.add
+ ;; CHECK-NEXT:      (i32.const 1)
+ ;; CHECK-NEXT:      (i32.const 2)
+ ;; CHECK-NEXT:     )
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:    (nop)
+ ;; CHECK-NEXT:    (local.get $scratch)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (i32.const 3)
+ ;; CHECK-NEXT:    (i32.const 4)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $add-twice-stacky (type $ret2)
+  i32.const 1
+  i32.const 2
+  i32.add
+  nop
+  i32.const 3
+  i32.const 4
+  i32.add
+ )
+
+ ;; CHECK:      (func $add-twice-stacky-2 (type $ret2) (result i32 i32)
+ ;; CHECK-NEXT:  (local $scratch i32)
+ ;; CHECK-NEXT:  (tuple.make
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (i32.const 1)
+ ;; CHECK-NEXT:    (i32.const 2)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (block (result i32)
+ ;; CHECK-NEXT:    (local.set $scratch
+ ;; CHECK-NEXT:     (i32.add
+ ;; CHECK-NEXT:      (i32.const 3)
+ ;; CHECK-NEXT:      (i32.const 4)
+ ;; CHECK-NEXT:     )
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:    (nop)
+ ;; CHECK-NEXT:    (local.get $scratch)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $add-twice-stacky-2 (type $ret2)
+  i32.const 1
+  i32.const 2
+  i32.add
+  i32.const 3
+  i32.const 4
+  i32.add
+  nop
+ )
+
+ ;; CHECK:      (func $add-twice-unreachable (type $ret2) (result i32 i32)
+ ;; CHECK-NEXT:  (tuple.make
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (unreachable)
+ ;; CHECK-NEXT:    (i32.const 2)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (i32.const 3)
+ ;; CHECK-NEXT:    (i32.const 4)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $add-twice-unreachable (type $ret2)
+  unreachable
+  i32.const 2
+  i32.add
+  i32.const 3
+  i32.const 4
+  i32.add
+ )
+
+ ;; CHECK:      (func $add-twice-unreachable-2 (type $ret2) (result i32 i32)
+ ;; CHECK-NEXT:  (drop
+ ;; CHECK-NEXT:   (i32.add
+ ;; CHECK-NEXT:    (i32.const 1)
+ ;; CHECK-NEXT:    (i32.const 2)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (tuple.make
```

Comment:
Why does this make a tuple? A block would be simpler (in particular I'm not sure how the optimizer handles tuples).

Reply:
This is a multivalue function, so it has to return a tuple.

The optimizer (along with everything else) handles tuples very poorly!

Reply:
I missed the function signature... makes sense now.